iter/iter.go

```diff
@@ -57,29 +58,69 @@ func ForEachIdx[T any](input []T, f func(int, *T)) { Iterator[T]{}.ForEachIdx(in
 // ForEachIdx is the same as ForEach except it also provides the
 // index of the element to the callback.
 func (iter Iterator[T]) ForEachIdx(input []T, f func(int, *T)) {
+	_ = iter.ForEachIdxCtx(context.Background(), input, func(_ context.Context, idx int, input *T) error {
+		f(idx, input)
+		return nil
+	})
+}
+
+// ForEachCtx is the same as ForEach except it also accepts a context
+// that it uses to manages the execution of tasks.
+// The context is cancelled on task failure and the first error is returned.
+func ForEachCtx[T any](octx context.Context, input []T, f func(context.Context, *T) error) error {
+	return Iterator[T]{}.ForEachCtx(octx, input, f)
+}
+
+// ForEachCtx is the same as ForEach except it also accepts a context
+// that it uses to manages the execution of tasks.
+// The context is cancelled on task failure and the first error is returned.
+func (iter Iterator[T]) ForEachCtx(octx context.Context, input []T, f func(context.Context, *T) error) error {
+	return iter.ForEachIdxCtx(octx, input, func(ctx context.Context, _ int, input *T) error {
+		return f(ctx, input)
+	})
+}
+
+// ForEachIdxCtx is the same as ForEachIdx except it also accepts a context
+// that it uses to manages the execution of tasks.
+// The context is cancelled on task failure and the first error is returned.
+func ForEachIdxCtx[T any](octx context.Context, input []T, f func(context.Context, int, *T) error) error {
+	return Iterator[T]{}.ForEachIdxCtx(octx, input, f)
+}
+
+// ForEachIdxCtx is the same as ForEachIdx except it also accepts a context
+// that it uses to manages the execution of tasks.
+// The context is cancelled on task failure and the first error is returned.
+func (iter Iterator[T]) ForEachIdxCtx(octx context.Context, input []T, f func(context.Context, int, *T) error) error {
 	if iter.MaxGoroutines == 0 {
 		// iter is a value receiver and is hence safe to mutate
 		iter.MaxGoroutines = defaultMaxGoroutines()
 	}
 
 	numInput := len(input)
-	if iter.MaxGoroutines > numInput {
+	if iter.MaxGoroutines > numInput && numInput > 0 {
 		// No more concurrent tasks than the number of input items.
 		iter.MaxGoroutines = numInput
 	}
 
 	var idx atomic.Int64
 	// Create the task outside the loop to avoid extra closure allocations.
-	task := func() {
+	task := func(ctx context.Context) error {
 		i := int(idx.Add(1) - 1)
-		for ; i < numInput; i = int(idx.Add(1) - 1) {
-			f(i, &input[i])
+		for ; i < numInput && ctx.Err() == nil; i = int(idx.Add(1) - 1) {
+			if err := f(ctx, i, &input[i]); err != nil {
+				return err
+			}
 		}
+		return ctx.Err() // nil if the context was never cancelled
 	}
 
-	var wg conc.WaitGroup
+	runner := pool.New().
+		WithContext(octx).
+		WithCancelOnError().
+		WithFirstError().
+		WithMaxGoroutines(iter.MaxGoroutines)
```

Comment:
I would prefer to not add a dependency on `pool` in the `iter` package. Because the iter package knows the number of inputs and the number of outputs in advance, it can be considerably more efficient than the `pool` package, which must work for an unbounded number of iterations.

I think it would be good to reconcile this PR with the patterns in https://github.com/sourcegraph/conc/pull/104. In particular, the `FailFast` flag should mean similar things iterating with `ForEachErr` and `ForEachCtx`.

Reply:
The abstractions built into `ContextPool` offered a simple way to achieve the functionality I was looking for with regard to context cancellation without having to duplicate code. Both `iter` and `pool` used `waitgroup` underneath. I'm not too sure what efficiency gains are made by having two separate implementations but I'd be happy to switch it back to original setup.

Additionally, as a further argument for using `ContextPool`, the `FailFast` behavior seems like it could be easily implemented by using the new bool to optionally call `WithFirstError()` on the underlying pool. Right now I call that by default assuming the caller is only interested in the first error when using a context.

I could add the *Err variants as extensions to what I have here that just return an error but don't require a context.